fontforge/dumppfa.c

```diff
@@ -977,16 +977,30 @@ void SC_PSDump(void (*dumpchar)(int ch,void *data), void *data,
     int i, last, first;
     SplineSet *temp;
 
-    if ( sc==NULL )
-return;
+    if ( sc==NULL ) return;
+
     last = first = layer;
     if ( layer==ly_all )
 	first = last = ly_fore;
     if ( sc->parent->multilayer ) {
 	first = ly_fore;
 	last = sc->layer_cnt-1;
     }
+
     for ( i=first; i<=last; ++i ) {
+
+    // Before we do anything, convert references to glyphs between 0x300 and
+    // 0x345 to splines. This is because Type1 fonts don't support references
+    // to them—or at least didn't in 2001—this format is ancient and retained
+    // primarily for backwards compatibility.
+    if ( sc->layers[i].refs!=NULL ) {
+        for ( ref = sc->layers[i].refs; ref!=NULL; ref=ref->next ) {
+            if (ref->unicode_enc >= 0x300 || ref->unicode_enc <= 0x345) {
+                SCRefToSplines(sc, ref, i);
```

Comment:
So: Starting a separate, totally independent code review conversation here ...

If this works at all it's because `SCRefToSplines()` modifies the `sc` argument, because there's no other way for it to have an effect.  Does this function somehow always get called with a copy of the font? Because otherwise exporting and then saving is going to unlink references, which is at least counter-intuitive (and probably just "wrong").

Reply:
(Added: Happily, there's already a code "model" of how to accomplish what's needed below in the `refs_to_splines` stanza. It's not exactly intuitive, but it's there.)

Reply:
@skef SCRefToSplines modifies `sc`, the code after this then uses the modified `sc` to write out the font.

Reply:
Right, but `sc` is (unless a copy is being made) a data structure maintained by FontForge as the in-memory representation of the glyph. So modifying `sc` is modifying the internal font representation.

Reply:
Oh, OK, I see what you mean. I can confirm that it is called with a copy of the font; when I save a PFA it doesn't cause the reference to be dissolved in my open SFD, no.

Reply:
If you want though @skef, to make this completely bulletproof since it's called in multiple places, I can change the code so everything works on a copy of `sc` and not the original. Let me know. :-)

Reply:
Oh: `if (ref->unicode_enc >= 0x300 || ref->unicode_enc <= 0x345) {` should be `if (ref->unicode_enc >= 0x300 && ref->unicode_enc <= 0x345) {`, right?

Reply:
If you were to:

1. Change the top guard to be `if ( sc->layers[i].refs!=NULL && !refs_to_splines ) {`
2. Confirm that every call path for the function that doesn't set `refs_to_splines` to true calls with a copy of `sc` (or an `sc` that will always be discarded for some other reason).
3. Add a comment about the need to call with a copy.

Then I think your approach would be fine. There aren't that many callers, but this function is used (for example) to export Encapsulated PostScript glyphs (with `refs_to_splines` set to true), and I don't know if it uses a copy of `sc` in that case. 

However, as I indicated somewhere else, it might be most "elegant" to invert the `refs_to_splines` clause in the later part of the function. Since the code is "right there" I guess my recommendation would be to try fiddling with that for a half-hour or so, and see if it works straightforwardly.

To be clearer, I'm recommending eliminating your loop and changing what now looks like:
```
            if ( refs_to_splines ) {
                if ( !pdfopers || !sc->parent->multilayer ) {
                    /* In PostScript, patterns are transformed by the page's */
                    /*  transformation matrix. In PDF they are not. */
                    /* Of course if we have no patterns we can still use this code */
                    for ( ref = sc->layers[i].refs; ref!=NULL; ref=ref->next ) {
                        A;
                    }
                } else {
                    /* If we get here we are outputting pdf, are type3, refs_2_splines */
/*  We need different gradients and patterns for different transform */
/*  matrices of references */
                    int j;
                    for ( ref = sc->layers[i].refs; ref!=NULL; ref=ref->next ) {
                        B;
                    }
                }
            } else {
                dumpstr(dumpchar,data,"    pop -1\n" );
                for ( ref = sc->layers[i].refs; ref!=NULL; ref=ref->next ) {
                    C;
                }
            }

```
to something like:
```
    int ref_to_dump = 0;

            for ( ref = sc->layers[i].refs; ref!=NULL; ref=ref->next ) {
                int j;
                if ( refs_to_splines || (ref->unicode_enc >= 0x300 && ref->unicode_enc <= 0x345) ) {
                    if ( !pdfopers || !sc->parent->multilayer ) {
                        /* In PostScript, patterns are transformed by the page's */
                        /*  transformation matrix. In PDF they are not. */
                        /* Of course if we have no patterns we can still use this code */
                        A;
                    } else {
                        /* We are outputting pdf, are type3, converting a ref to splines */
                        /* We need different gradients and patterns for different transform */
                        /* matrices of references */
                        B;
                    }
                } else
                    ref_to_dump = 1;
            }
            if ( ref_to_dump ) {
                dumpstr(dumpchar,data,"    pop -1\n" );
                for ( ref = sc->layers[i].refs; ref!=NULL; ref=ref->next ) {
                    if ( ref->unicode_enc >= 0x300 && ref->unicode_enc <= 0x345 )
                        continue;
                    C;
                }
            }
```
Now **this will reorder**. But given that all the splines are output first anyway, I think it will reorder in the same way your modification does (in that it converts certain references to splines).

Reply:
I believe we're heading towards eliminating this code, so I'm resolving this part of the discussion.

Reply:
Per our discussion I've just decided to revert `dumppfa.c` to its state in master.